### PR TITLE
EASY-2537: upload files to data/original/*

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-2537 upload to original

#### When applied it will
* [x] merge `StagedFilesTarget` into `DataFiles`
* [x] make compatible with existing datasets without the origin folder
* 

#### How should this be manually tested?

* [x] create two bags and upload attached zip
* delete files outside the folder "original" from the second bag
* build and deploy
* upload another file to the both bags
* create another bag, upload something
* submit the three bags
* check the result in the files tab of the archived datasets

after deployment, the folder "original" will still show up in the first draft but no longer in the second.

image

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 1]
```diff
@@ -90,12 +92,10 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
...
     val absolutePath = bag.data / path.toString
     val fileExists = manifests.get.contains(absolutePath)
-    if (fileExists) {
-      val checksum = manifests get absolutePath
-      Success(FileInfo(path.getFileName.toString, bag.data.relativize(absolutePath.parent), checksum))
-    }
+    if (fileExists) Success(toFileInfo(absolutePath, checksum = manifests get absolutePath))
     else {
-      Failure(NoSuchFileInDepositException(absolutePath, path))
+      val path1 = uploadRoot.relativize(absolutePath)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: inline into the next line

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 2]
```diff
@@ -124,8 +124,10 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
    */
   def delete(path: Path): Try[Unit] = {
     val file = bag.data / path.toString
-    if (!file.exists) Failure(NoSuchFileInDepositException(file, path))
-    else (if (file.isDirectory) removeDir(file.walk().toStream)
+    if (!file.exists) {
+      val relPath = uploadRoot.relativize(file)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: inline into next line

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 3]
```diff
@@ -124,8 +124,10 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
...
+      Failure(NoSuchFileInDepositException(relPath))
+    } else (if (file.isDirectory) removeDir(file.walk().toStream)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: can't this be an `else if`?

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala [hunk 4]
```diff
@@ -162,8 +162,7 @@ class DepositServlet(app: EasyDepositApiApp)
     {
       for {
         uuid <- getUUID
-        path <- getPath
-        _ = logger.info(s"[$uuid] upload file to path '${ path.toString.toOption.getOrElse("/") }'")
+        path <- getRelativeLocationInBagData(s"[$uuid] upload files to path") // plural)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]:
```suggestion
        path <- getRelativeLocationInBagData(s"[$uuid] upload files to path") // plural
```

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala [hunk 5]
```diff
@@ -232,3 +234,6 @@ class DepositServlet(app: EasyDepositApiApp)
     }
   }
 }
+object DepositServlet {
+  val uploadRoot = "original"
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: is there a better place for this constant?

File: src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala [hunk 6]
```diff
@@ -101,7 +101,7 @@ class IntegrationSpec extends TestSupportFixture with ServletFixture with Scalat
     val uuid = createDeposit
 
     // upload files in a folder (more variations in UploadSpec)
-    val dataFilesBase = DepositDir(testDir / "drafts", "foo", UUID.fromString(uuid)).getDataFiles.get.bag.data
+    val dataFilesBase = DepositDir(testDir / "drafts", "foo", UUID.fromString(uuid)).getDataFiles.get.bag.data / DepositServlet.uploadRoot
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala]: In this test you use the `uploadRoot` constant, whereas in other places you use `"original"` as a `String`. Please make sure that is consistent.

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala (hunk 5 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: see commit comments

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 7]
```diff
@@ -107,11 +156,12 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
    * @return `true` if a new file was created, `false` if an existing file was overwritten
    */
   def write(is: InputStream, path: Path): Try[Boolean] = {
-    val fileExists = (bag.data / path.toString).exists
+    val fileExists = (uploadRoot / path.toString).exists
+    val pathInDataDir = bag.data.relativize(uploadRoot / path.toString)
     for {
-      _ <- if (fileExists) bag.removePayloadFile(path)
+      _ <- if (fileExists) cleanUp(pathInDataDir) // TODO see fetchFiles in StagedFilesTarget
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: what does this TODO mean? I don't know what `StagedFilesTarget` is, neither can I find any other reference to it

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 8]
```diff
@@ -123,11 +173,15 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
...
-    if (!file.exists) Failure(NoSuchFileInDepositException(file, path))
-    else (if (file.isDirectory) removeDir(file.walk().toStream)
-          else bag.removePayloadFile(path)
-      ).flatMap(_.save)
+    val file = uploadRoot / path.toString
+    if (!file.exists) {
+      Failure(NoSuchFileInDepositException(path))
+    }
+    else {
+      val triedBag = if (file.isDirectory) removeDir(file.walk().toStream)
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: IMO `listRecursively` is more descriptive than `walk`, even though they do the same thing.

```suggestion
      val triedBag = if (file.isDirectory) removeDir(file.listRecursively().toStream)
```

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala (hunk 7 above)
Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: ...somewhat later...
O wait, `StagedFilesTarget` is a file you deleted... Perhaps the TODO is no longer relevant?

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala [hunk 9]
```diff
@@ -36,16 +35,7 @@ class DepositServlet(app: EasyDepositApiApp)
   extends ProtectedServlet(app)
     with FileUploadSupport {
 
-  configureMultipartHandling(app.multipartConfig.copy(location = app.multipartConfig.location
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: is this whole check no longer necessary?

File: src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala [hunk 10]
```diff
@@ -147,6 +145,30 @@ class DataFilesSpec extends TestSupportFixture {
...
+    dataFiles.list(Paths.get("original")) shouldBe Success(Seq.empty)
+  }
+
+  it should " return an empty list on foo/bar of a new bag" in {
+    val bag = DansV0Bag
+      .empty(testDir / "testBag").getOrRecover(fail("could not create test bag", _))
+    bag.save()
+    val dataFiles = DataFiles(bag)
+    dataFiles.list(Paths.get("foo/bar")) shouldBe Success(Seq.empty)
+  }
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala]: please remove the leading spaces in the test descriptions

File: src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala [hunk 11]
```diff
@@ -147,6 +145,30 @@ class DataFilesSpec extends TestSupportFixture {
     ))
   }
 
+  it should " return an empty list on a new bag" in {
+    val bag = DansV0Bag
+      .empty(testDir / "testBag").getOrRecover(fail("could not create test bag", _))
+    bag.save()
+    val dataFiles = DataFiles(bag)
+    dataFiles.list() shouldBe Success(Seq.empty)
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala]: But does an empty bag also **_create_** the `original` directory? If not, the two tests below are identical.

File: src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala [hunk 12]
```diff
@@ -27,7 +27,6 @@ import nl.knaw.dans.easy.deposit.Errors.NoSuchFileInDepositException
 import nl.knaw.dans.easy.deposit.docs.FileInfo
 import nl.knaw.dans.lib.error._
 
-import scala.collection.JavaConverters._
 import scala.util.{ Failure, Success }
 
 class DataFilesSpec extends TestSupportFixture {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala]: Not entirely sure, but looking through this class, I don't see any tests for the backwards compatibility cases...

File: src/test/scala/nl.knaw.dans.easy.deposit/StagedFilesSpec.scala [hunk 13]
```diff
@@ -17,26 +17,25 @@ package nl.knaw.dans.easy.deposit
...
+  private val stagedDir = testDir / "staged"
 
   override def beforeEach(): Unit = {
     super.beforeEach()
     clearTestDir()
     stagedDir.createDirectories()
   }
 
-  "moveAllFrom" should "add payload files" in {
+  "DataFiles.moveAll" should "add payload files" in {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.deposit/StagedFilesSpec.scala]: these tests should be moved to `DataFilesSpec` to mimic the class structure in `src/main`

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala (hunk 9 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: The app class does a same-mount check by creating and moving files. That makes these checks redundant.

File: src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala (hunk 11 above)
Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.deposit/DataFilesSpec.scala]: `DansV0Bag` has no knowledge about the original folder and does not support empty directories.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 14]
```diff
@@ -36,11 +37,11 @@ import scala.util.{ Failure, Success, Try }
  *
  * @param bag the bag containing the data files
  */
-case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
+case class DataFiles(bag: DansBag, bagId: UUID) extends DebugEnhancedLogging {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]:
```suggestion
case class DataFiles(bag: DansBag, depositId: UUID) extends DebugEnhancedLogging {
```

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 15]
```diff
@@ -35,16 +37,66 @@ import scala.util.{ Failure, Success, Try }
...
+  private lazy val uploadRoot = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: Would you please add a comment briefly explaining the reason for having both these options: a.k.a. old situation vs. new situation. Future generations of developers will thank you later 😉

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 16]
```diff
@@ -36,16 +37,64 @@ import scala.util.{ Failure, Success, Try }
...
+  }
+
+  // though currently we don't create fetch files, let us not run into trouble whenever we do
+  private lazy val fetchFiles = bag.fetchFiles
+    .map(fetchFile =>
+      bag.data.relativize(fetchFile.file)
+    )
+
+  def cleanUp(bagRelativePath: Path): Try[Any] = {
+    val absFile = bag.data / bagRelativePath.toString
```
[the remark this answers is not in the document]
Comment (rvanheest) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: Perhaps you can add a comment for this? Just to make sure we don't stumble over this in the future again.

[the remark this answers is not in the document]
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: I payed so much attention to consistent naming conventions...

[the remark this answers is not in the document]
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: The argument is now called `bagPayloadPath` that should make it obvious it's relative to bag/data

**Pull request merged**